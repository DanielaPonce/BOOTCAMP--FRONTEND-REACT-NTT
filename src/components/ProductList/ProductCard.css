
.product-card {
	background-color: var(--white);
	border-radius: 10px;
	box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
	overflow: hidden;
	width: 300px;
	transition: transform 0.3s;

	position: relative;
}

.product-card:hover {
	transform: translateY(-16px);
}

.product-card img {
	width: 100%;
	height: 200px;
	object-fit: contain;
}

.product-details {
	padding: 1rem;
	text-align: center;
}

.product-details h3 {
	margin: 0.5rem 0;
}

.product-details p {
	font-size: 0.9rem;
	color: var(--text-color);
}

.product-details .category {
	font-weight: bold;
	color: var(--secondary-color);
	background-color: var(--white);
	padding: 4px 7px;
	border-radius: 4px;
	display: inline-block;
	margin-bottom: 0.5rem;

	position: absolute;
	z-index: 1;
	top: 2%;
	left: 2%;
}

.product-details .price {
	font-size: 1.2rem;
	color: var(--text-color);
	margin: 0.5rem 0;
}

.product-details button {
	background-color: var(--secondary-color);
	color: var(--white);
	padding: 0.7rem 1.5rem;
	border: none;
	border-radius: 5px;
	cursor: pointer;
	transition: background-color 0.3s;
}

.product-details button:hover {
	background-color: var(--hover-button-color);
}
