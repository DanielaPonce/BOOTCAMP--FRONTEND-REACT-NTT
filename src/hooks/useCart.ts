import { useEffect } from 'react';
import { useCartContext } from '../context/CartContext';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

export const useCart = () => {
	const { state, dispatch } = useCartContext();

	useEffect(() => {
		localStorage.setItem(
			LocalStorageConstants.CartKey,
			JSON.stringify(state)
		);
	}, [state]);

	const addToCart = (product: Product) =>
		dispatch({ type: Actions.ADD_TO_CART_TYPE, payload: product });

	const removeFromCart = (id: number) =>
		dispatch({ type: Actions.REMOVE_FROM_CART_TYPE, payload: id });

	const updateQuantity = (id: number, quantity: number) =>
		dispatch({
			type: Actions.UPDATE_QUANTITY_TYPE,
			payload: { id, quantity }
		});

	const clearCart = () => dispatch({ type: Actions.CLEAR_CART_TYPE });

	return {
		cart: state.cart,
		total: state.total,
		itemCount: state.itemCount,
		addToCart,
		removeFromCart,
		updateQuantity,
		clearCart
	};
};
