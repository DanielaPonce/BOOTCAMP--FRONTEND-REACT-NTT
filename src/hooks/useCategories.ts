import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
