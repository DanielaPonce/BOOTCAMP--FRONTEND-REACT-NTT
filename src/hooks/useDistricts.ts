import { useEffect, useState } from 'react';

interface District {
	id: number;
	name: string;
}

export const useDistricts = () => {
	const [districts, setDistricts] = useState<District[]>([]);

	useEffect(() => {
		const fetchDistricts = async () => {
			const response = await fetch('/src/assets/mock/districts.json');
			const data: District[] = await response.json();
			setDistricts(data);
		};

		fetchDistricts();
	}, []);

	return districts;
};
