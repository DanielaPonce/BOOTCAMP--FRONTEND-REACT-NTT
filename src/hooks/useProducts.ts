import { useEffect, useState } from 'react';
import { fetchAllProducts, fetchProductsByCategory } from '../api/product.api';
import { Product } from '../models/product.types';
import { Constants } from '../utils/constants';

export const useProducts = (category: string, search: string) => {
	const [allProducts, setAllProducts] = useState<Product[]>([]);
	const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [currentPage, setCurrentPage] = useState(1);

	const itemsPerPage = 8;

	const loadProducts = async () => {
		setIsLoading(true);
		try {
			let products: Product[] = [];
			if (category === Constants.defaultSelectValue) {
				products = await fetchAllProducts();
			} else {
				products = await fetchProductsByCategory(category);
			}
			setAllProducts(products);
			setFilteredProducts(products);
			setCurrentPage(1);
		} catch (error) {
			console.log('Error al cargar los productos:', error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		loadProducts();
	}, [category]);

	useEffect(() => {
		const filtered = allProducts.filter((product) =>
			product.title.toLowerCase().includes(search.toLowerCase())
		);
		setFilteredProducts(filtered);
		setCurrentPage(1);
	}, [search, allProducts]);

	const paginatedProducts = filteredProducts.slice(
		(currentPage - 1) * itemsPerPage,
		currentPage * itemsPerPage
	);

	const totalPages = Math.ceil(filteredProducts.length / itemsPerPage);

	return {
		products: paginatedProducts,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination: totalPages !== 1
	};
};
