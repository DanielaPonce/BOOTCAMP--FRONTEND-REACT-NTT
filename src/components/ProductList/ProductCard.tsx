import { FC } from 'react';
import { Product } from '../../models/product.types';
import { useCart } from '../../hooks/useCart';

const ProductCard: FC<Product> = (props) => {
	const { id, title, description, price, thumbnail, category } = props;

	const { addToCart } = useCart();

	const handleAddToCart = () => {
		addToCart(props);
	};

	return (
		<div className="product-card">
			<img src={thumbnail} alt={title} />
			<div className="product-details">
				<h3>{title}</h3>
				<p>{description}</p>
				<span className="category">{category}</span>
				<div className="price">Precio: S/ {price}</div>
				<button onClick={handleAddToCart}>Agregar al carrito</button>
			</div>
		</div>
	);
};

export default ProductCard;
