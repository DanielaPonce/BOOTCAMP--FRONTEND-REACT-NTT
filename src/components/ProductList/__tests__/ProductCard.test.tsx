import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { jest } from '@jest/globals';
import ProductCard from '../ProductCard';
import { useCart } from '../../../hooks/useCart';

jest.mock('../../../hooks/useCart');

describe('ProductCard Component', () => {
	const mockAddToCart = jest.fn();
	const mockProduct = {
		id: 1,
		title: 'Producto de prueba',
		description: 'Descripción del producto de prueba',
		price: 99.99,
		thumbnail: 'https://via.placeholder.com/150',
		category: 'Categoría de prueba'
	};

	beforeEach(() => {
		jest.clearAllMocks();
		(useCart as jest.Mock).mockReturnValue({ addToCart: mockAddToCart });
	});

	test('renders correctly', () => {
		render(<ProductCard {...mockProduct} />);

		const image = screen.getByAltText(mockProduct.title);
		expect(image).toBeInTheDocument();
		expect(image).toHaveAttribute('src', mockProduct.thumbnail);

		expect(screen.getByText(mockProduct.title)).toBeInTheDocument();
		expect(screen.getByText(mockProduct.description)).toBeInTheDocument();
		expect(
			screen.getByText(`Precio: S/ ${mockProduct.price}`)
		).toBeInTheDocument();
		expect(screen.getByText(mockProduct.category)).toBeInTheDocument();

		const button = screen.getByRole('button', {
			name: /Agregar al carrito/i
		});
		expect(button).toBeInTheDocument();
	});

	test('calls addToCart with correct product when button is clicked', () => {
		render(<ProductCard {...mockProduct} />);

		const button = screen.getByRole('button', {
			name: /Agregar al carrito/i
		});
		fireEvent.click(button);

		expect(mockAddToCart).toHaveBeenCalledTimes(1);
		expect(mockAddToCart).toHaveBeenCalledWith(mockProduct);
	});

	test('matches snapshot', () => {
		const { asFragment } = render(<ProductCard {...mockProduct} />);
		expect(asFragment()).toMatchSnapshot();
	});
});
