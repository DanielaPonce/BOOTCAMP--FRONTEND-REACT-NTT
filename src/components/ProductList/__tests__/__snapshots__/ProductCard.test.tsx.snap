// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <img
      alt="Producto de prueba"
      src="https://via.placeholder.com/150"
    />
    <div
      class="product-details"
    >
      <h3>
        Producto de prueba
      </h3>
      <p>
        Descripción del producto de prueba
      </p>
      <span
        class="category"
      >
        Categoría de prueba
      </span>
      <div
        class="price"
      >
        Precio: S/ 99.99
      </div>
      <button>
        Agregar al carrito
      </button>
    </div>
  </div>
</DocumentFragment>
`;
